aten/src/ATen/native/vulkan/Vulkan.h

```diff
@@ -451,6 +463,40 @@ class ComputeUnit final {
   VkShaderModule computeShaderModule_;
 };
 
+class ComputeUnitFactory {
+ public:
+  ComputeUnitFactory(const VkDevice& device);
+  ~ComputeUnitFactory();
+  ComputeUnitFactory(const ComputeUnitFactory&) = default;
+  ComputeUnitFactory& operator=(const ComputeUnitFactory&) = default;
+  ComputeUnitFactory(ComputeUnitFactory&&) = default;
+  ComputeUnitFactory& operator=(ComputeUnitFactory&&) = default;
+
+#ifdef USE_VULKAN_SHADERC_RUNTIME
+  ComputeUnit& get(
+      const char* key,
+      const char* glslSrc,
+      const VkDescriptorSetLayout& descrSetLayout,
+      WorkGroupSize workGroupSize);
+#else
+  ComputeUnit& get(
+      const char* key,
+      const uint32_t* code,
+      const uint32_t codeSize,
+      const VkDescriptorSetLayout& descrSetLayout,
+      WorkGroupSize& workGroupSize);
+#endif
+ private:
+  std::string getCacheKey(const char* key, WorkGroupSize& workGroupSize);
+  ComputeUnit& get(
+      const std::string& cacheKey,
+      std::function<std::shared_ptr<ComputeUnit>()> factoryFn);
+
+  VkDevice device_;
+  VkPipelineCache pipelineCache_;
+  std::map<std::string, std::shared_ptr<ComputeUnit>> computeUnits_;
```

Comment:
Consider using a std::unordered_map (aka hash map) instead of a std::map.  std::map is a poor data structure in general for performance: O(log(n)) look up and poor cache performance due to scattered memory accesses which is even more detrimental to performance when dealing with few (say hundreds) of elements like we are here.  In those scenarios the best data structure is typically a vector with a good old boring linear search or for slightly bigger arrays (1000s of elements) sorted vector plus a binary search.  A hash map is perfectly fine for now, but just as an FYI if you wanted to go down the route of vector you would store [uint64_t, std::shared_ptr<ComputeUnit>] in the vector where uint64_t is the hash key, and you would do the lookup by using hash(string) as key. Collisions should be very rare and could be handled by storing identical keys inline and doing a linear search over them.

Chances are high this never ends up mattering in our usage here so unordered_map is perfectly fine.

Reply:
Thanks, I did not know the difference between std::map and std::unordered_map, looks like map works as HashTree (balanced binary tree) while unordered_map juts as a hash map without preserving ordering. I will change it to unordered_map.